pbsmetrics/prometheus/metrics.go

```diff
@@ -0,0 +1,374 @@
+package prometheusmetrics
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/pbsmetrics"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Metrics defines the Prometheus metrics backing the MetricsEngine implementation.
+type Metrics struct {
+	Registry *prometheus.Registry
+
+	// General Metrics
+	connectionsClosed            prometheus.Counter
+	connectionsError             *prometheus.CounterVec
+	connectionsOpened            prometheus.Counter
+	cookieSync                   prometheus.Counter
+	impressions                  *prometheus.CounterVec
+	impressionsLegacy            prometheus.Counter
+	prebidCacheWriteTimer        *prometheus.HistogramVec
+	requests                     *prometheus.CounterVec
+	requestsTimer                *prometheus.HistogramVec
+	requestsWithoutCookie        *prometheus.CounterVec
+	storedImpressionsCacheResult *prometheus.CounterVec
+	storedRequestCacheResult     *prometheus.CounterVec
+
+	// Adapter Metrics
+	adapterBids          *prometheus.CounterVec
+	adapterCookieSync    *prometheus.CounterVec
+	adapterErrors        *prometheus.CounterVec
+	adapterPanics        *prometheus.CounterVec
+	adapterPrices        *prometheus.HistogramVec
+	adapterRequests      *prometheus.CounterVec
+	adapterRequestsTimer *prometheus.HistogramVec
+	adapterUserSync      *prometheus.CounterVec
+
+	// Account Metrics
+	accountRequests *prometheus.CounterVec
+}
+
+const (
+	accountLabel         = "account"
+	actionLabel          = "action"
+	adapterLabel         = "adapter"
+	adapterErrorLabel    = "adapter_error"
+	bidTypeLabel         = "bid_type"
+	cacheResultLabel     = "cache_result"
+	connectionErrorLabel = "connection_error"
+	hasBidsLabel         = "has_bids"
+	hasCookieLabel       = "has_cookie"
+	isAudioLabel         = "audio"
+	isBannerLabel        = "banner"
+	isNativeLabel        = "native"
+	isVideoLabel         = "video"
+	markupDeliveryLabel  = "delivery"
+	privacyBlockedLabel  = "privacy_blocked"
+	requestStatusLabel   = "request_status"
+	requestTypeLabel     = "request_type"
+	successLabel         = "success"
+)
+
+const (
+	connectionAcceptError = "accept"
+	connectionCloseError  = "close"
+)
+
+const (
+	markupDeliveryAdm  = "adm"
+	markupDeliveryNurl = "nurl"
+)
+
+// NewMetrics initializes a new Prometheus metrics instance with preloaded label values.
+func NewMetrics(cfg config.PrometheusMetrics) *Metrics {
+	requestTimeBuckets := []float64{0.05, 0.1, 0.15, 0.20, 0.25, 0.3, 0.4, 0.5, 0.75, 1}
+	cacheWriteTimeBuckts := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
+	priceBuckets := []float64{250, 500, 750, 1000, 1500, 2000, 2500, 3000, 3500, 4000}
```

Comment:
A future PR will make this configurable. Hardcoding for now. Prometheus devs provide guidance of around 10 buckets per histogram. The values are chosen based on analysis of InfluxDB values. Hopefully I got the math right. :)